lib/puppet/util/execution.rb

```diff
@@ -187,18 +187,30 @@ def self.execute(command, options = NoOptionsSpecified)
     null_file = Puppet.features.microsoft_windows? ? 'NUL' : '/dev/null'
 
     begin
+      reader, writer = IO.pipe unless options[:squelch]
       stdin = Puppet::FileSystem.open(options[:stdinfile] || null_file, nil, 'r')
-      stdout = options[:squelch] ? Puppet::FileSystem.open(null_file, nil, 'w') : Puppet::FileSystem::Uniquefile.new('puppet')
+      stdout = options[:squelch] ? Puppet::FileSystem.open(null_file, nil, 'w') : writer
       stderr = options[:combine] ? stdout : Puppet::FileSystem.open(null_file, nil, 'w')
 
       exec_args = [command, options, stdin, stdout, stderr]
+      output = ''
+      read_succeeded = false
 
       if execution_stub = Puppet::Util::ExecutionStub.current_value
-        return execution_stub.call(*exec_args)
+        child_pid = execution_stub.call(*exec_args)
+        [stdin, stdout, stderr].each {|io| io.close rescue nil}
```

Comment:
We already do close them in Puppet Server, although it seems like it would have better if we didn't have to.  See [this code](https://github.com/puppetlabs/puppetserver/blob/2.7.2/src/ruby/puppetserver-lib/puppet/server/execution.rb#L8-L12).

If we add this here, we might get exceptions when the second set of close calls are made.  That should be okay with the swallowing `rescue`, though, right?  Not sure if anything would be written to the log when this happens?